Add decomp for pad_sequence

zhxchen17 opened the pull request:
Summary: currently pad_sequence caused symbolic shape specialization in export which is unintended. Adding a decomp seems to work to avoid the c++ kernel which caused the specialization.

Test Plan: buck test mode/opt caffe2/test:test_export -- -r pad_sequence

Reviewed By: SherlockNoMad

Differential Revision: D52345667

Skylion007 commented on torch/_decomp/decompositions.py:
```diff
@@ -2284,6 +2284,32 @@ def _index_add(
+    sequences_size = len(sequences)
+    max_size = sequences[0].size()
+    trailing_dims = max_size[1:]
+    max_len = max([x.size(0) for x in sequences])
```
```suggestion
    max_len = max(x.size(0) for x in sequences)
```

lezcano commented on torch/_decomp/decompositions.py:
```diff
@@ -2284,6 +2284,32 @@ def _index_add(
+        currseq = sequences[i]
+        length_i = currseq.size(0)
+        if batch_first:
+            out[i, :length_i].copy_(currseq)
```
in-place ops in decompositions are not kosher. The way to do this is to figure out by how much you need to pad every tensor, and concatenate a `new_full` of the right length to every input. At the end, you just stack all of them together.

peterbell10 replied:
One thing to note, I tried both and inductor actually does a much better with the copies... 1 kernel vs `sequences_size` kernels. 

I think the kosher way to write this would be
```
row = aten.constant_pad_nd(curseq, (0, row_size - curseq.size(0)), pad_value)
out = aten.select_scatter(out, row, dim=dim, i)
```

Of course in an ideal world `torch.cat` would be able to code that's just as good.

lezcano replied:
I was going to say that the generated code was going to look horrible, but note that this PR doesn't add this decomposition to core so that's alright.

I was thinking about doing something with `pad` but didn't post it either. Agree this is a much better way of doing this. We should try write some passes to go from full + cat -> pad and, if it generates better code, from stack into a bunch of select scatter (or just generate better code for cat directly...)

lezcano reviewed (commented):
Can you add tests for `batch_first=True`? Can you also add tests with several inputs?

zhxchen17 committed:
Add decomp for pad_sequence (#116285)

Summary:

currently pad_sequence caused symbolic shape specialization in export which is unintended. Adding a decomp seems to work to avoid the c++ kernel which caused the specialization.

Test Plan: buck test mode/opt caffe2/test:test_export -- -r pad_sequence

Reviewed By: SherlockNoMad

Differential Revision: D52345667
